binapigen/run.go

```diff
@@ -16,94 +16,151 @@ package binapigen
 }
 
-func Run(apiDir string, filesToGenerate []string, opts Options, f func(*Generator) error) {
-	if err := run(apiDir, filesToGenerate, opts, f); err != nil {
+func ResolveVppInput(input string) (*VppInput, error) {
+	vppInput := &VppInput{}
+
+	if input == "" {
+		input = vppapi.DefaultDir
+	}
+
+	u, err := url.Parse(input)
+	if err != nil {
+		logrus.Warnf("parsing url error: %v", err)
+	} else {
+		switch u.Scheme {
+		case "", "file":
+			i, err := os.Stat(input)
+			if err != nil {
+				return nil, fmt.Errorf("file error: %v", err)
+			} else {
+				if i.IsDir() {
+					apidir := vppapi.ResolveApiDir(u.Path)
+					logrus.Debugf("path %q resolved to api dir: %v", u.Path, apidir)
+					apiFiles, err := vppapi.ParseDir(apidir)
+					if err != nil {
+						logrus.Warnf("vppapi parsedir error: %v", err)
+					} else {
+						vppInput.ApiFiles = apiFiles
+						logrus.Infof("resolved %d apifiles", len(apiFiles))
+					}
+					vppInput.VppVersion = vppapi.ResolveVPPVersion(u.Path)
+					if vppInput.VppVersion == "" {
+						vppInput.VppVersion = "unknown"
+					}
+				} else {
+					return nil, fmt.Errorf("files not supported")
+				}
+			}
+		case "http", "https":
+			return nil, fmt.Errorf("http(s) not yet supported")
+		case "git", "ssh":
+			return nil, fmt.Errorf("ssh/git not yet supported")
+		default:
+			return nil, fmt.Errorf("unsupported scheme: %v", u.Scheme)
+		}
+	}
+
+	return vppInput, nil
+}
+
+func Run(vppInput *VppInput, opts Options, f func(*Generator) error) {
+	if err := run(vppInput, opts, f); err != nil {
 		fmt.Fprintf(os.Stderr, "%s: %v\n", filepath.Base(os.Args[0]), err)
 		os.Exit(1)
 	}
 }
 
-func run(apiDir string, filesToGenerate []string, opts Options, fn func(*Generator) error) error {
-	apiFiles, err := vppapi.ParseDir(apiDir)
-	if err != nil {
-		return err
-	}
+func run(vppInput *VppInput, opts Options, genFn func(*Generator) error) error {
+	var err error
 
 	if opts.ImportPrefix == "" {
-		opts.ImportPrefix, err = resolveImportPath(opts.OutputDir)
+		opts.ImportPrefix, err = ResolveImportPath(opts.OutputDir)
 		if err != nil {
 			return fmt.Errorf("cannot resolve import path for output dir %s: %w", opts.OutputDir, err)
 		}
 		logrus.Debugf("resolved import path prefix: %s", opts.ImportPrefix)
 	}
 
-	gen, err := New(opts, apiFiles, filesToGenerate)
+	gen, err := New(opts, vppInput)
 	if err != nil {
 		return err
 	}
 
-	gen.vppVersion = vppapi.ResolveVPPVersion(apiDir)
-	if gen.vppVersion == "" {
-		gen.vppVersion = "unknown"
+	if genFn == nil {
+		genFn = GenerateDefault
 	}
-
-	if fn == nil {
-		GenerateDefault(gen)
-	} else {
-		if err := fn(gen); err != nil {
-			return err
-		}
+	if err := genFn(gen); err != nil {
+		return err
 	}
+
 	if err = gen.Generate(); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func GenerateDefault(gen *Generator) {
+func GeneratePlugins(genPlugins []string) func(*Generator) error {
+	return func(gen *Generator) error {
+		for _, file := range gen.Files {
```

Comment:
Maybe an API concern with this one, as we are consuming this with external plugins, 
It might be good to leave the choice to the plugin whether it wants to loop over the files or not.
i.e.  move the loop `for _, file := range gen.File` to the plugin themselves, as this would allow a plugin
to only run once knowing all files, or to have generation depending on groups of files, etc...